### PR TITLE
Refactor Operation Interface

### PR DESCRIPTION
**Pull request by JLLeitschuh**
- Operations now use local state variables for input and output sockets rather than regenerating them on the fly
- Moved name, description, category, icon, and aliases to a separate description object
- Socket classes are now interfaces
- Refactors Generated Operations to be in source code
- Changed socket get methods to return List<FooSocket> instead of FooSocket<?>[]

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -250,6 +250,7 @@ project(":core") {
         removeExisting true
     }
     generateCodeFromSource.onlyIf { !project.hasProperty('skipGenerate') }
+    //generateCodeFromSource.onlyIf { false }
```
Comment (SamCarlberg) [inline build.gradle]: Remove this line

File: core/src/main/java/edu/wpi/grip/core/operations/opencv/templated/FiveSourceOneDestinationOperation.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import com.google.common.collect.ImmutableList;
+import edu.wpi.grip.core.Operation;
+import edu.wpi.grip.core.sockets.InputSocket;
+import edu.wpi.grip.core.sockets.OutputSocket;
+import edu.wpi.grip.core.sockets.SocketHint;
+
+import java.util.List;
+
+
+public class FiveSourceOneDestinationOperation<T1, T2, T3, T4, T5, R> implements Operation {
```
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/opencv/templated/FiveSourceOneDestinationOperation.java]: Add documentation for how and when to use this

File: core/src/main/java/edu/wpi/grip/core/operations/opencv/templated/FourSourceOneDestinationOperation.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+import com.google.common.collect.ImmutableList;
+import edu.wpi.grip.core.Operation;
+import edu.wpi.grip.core.sockets.InputSocket;
+import edu.wpi.grip.core.sockets.OutputSocket;
+import edu.wpi.grip.core.sockets.SocketHint;
+
+import java.util.List;
+
+public class FourSourceOneDestinationOperation<T1, T2, T3, T4, R> implements Operation {
```
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/opencv/templated/FourSourceOneDestinationOperation.java]: Add documentation for how and when to use this

File: core/src/main/java/edu/wpi/grip/core/operations/opencv/templated/OneSourceOneDestinationOperation.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+import autovalue.shaded.com.google.common.common.collect.ImmutableList;
+import edu.wpi.grip.core.Operation;
+import edu.wpi.grip.core.sockets.InputSocket;
+import edu.wpi.grip.core.sockets.OutputSocket;
+import edu.wpi.grip.core.sockets.SocketHint;
+
+import java.util.List;
+
+public class OneSourceOneDestinationOperation<T1, R> implements Operation {
```
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/opencv/templated/OneSourceOneDestinationOperation.java]: Add documentation for how and when to use this

File: core/src/main/java/edu/wpi/grip/core/operations/opencv/templated/SevenSourceOneDestinationOperation.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+import com.google.common.collect.ImmutableList;
+import edu.wpi.grip.core.Operation;
+import edu.wpi.grip.core.sockets.InputSocket;
+import edu.wpi.grip.core.sockets.OutputSocket;
+import edu.wpi.grip.core.sockets.SocketHint;
+
+import java.util.List;
+
+public class SevenSourceOneDestinationOperation<T1, T2, T3, T4, T5, T6, T7, R> implements Operation {
```
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/opencv/templated/SevenSourceOneDestinationOperation.java]: Add documentation for how and when to use this

File: core/src/main/java/edu/wpi/grip/core/operations/opencv/templated/SixSourceOneDestinationOperation.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import com.google.common.collect.ImmutableList;
+import edu.wpi.grip.core.Operation;
+import edu.wpi.grip.core.sockets.InputSocket;
+import edu.wpi.grip.core.sockets.OutputSocket;
+import edu.wpi.grip.core.sockets.SocketHint;
+
+import java.util.List;
+
+
+public class SixSourceOneDestinationOperation<T1, T2, T3, T4, T5, T6, R> implements Operation {
```
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/opencv/templated/SixSourceOneDestinationOperation.java]: Add documentation for how and when to use this

File: core/src/main/java/edu/wpi/grip/core/operations/opencv/templated/SixSourceOneDestinationOperation.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        this.input2 = inputSocketFactory.create(t2SocketHint);
+        this.input3 = inputSocketFactory.create(t3SocketHint);
+        this.input4 = inputSocketFactory.create(t4SocketHint);
+        this.input5 = inputSocketFactory.create(t5SocketHint);
+        this.input6 = inputSocketFactory.create(t6SocketHint);
+        this.output = outputSocketFactory.create(rSocketHint);
+    }
+
+
+    public SixSourceOneDestinationOperation(
```
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/opencv/templated/SixSourceOneDestinationOperation.java]: Add documentation for how and when to use this

File: core/src/main/java/edu/wpi/grip/core/operations/opencv/templated/ThreeSourceOneDestinationOperation.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+import com.google.common.collect.ImmutableList;
+import edu.wpi.grip.core.Operation;
+import edu.wpi.grip.core.sockets.InputSocket;
+import edu.wpi.grip.core.sockets.OutputSocket;
+import edu.wpi.grip.core.sockets.SocketHint;
+
+import java.util.List;
+
+public class ThreeSourceOneDestinationOperation<T1, T2, T3, R> implements Operation {
```
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/opencv/templated/ThreeSourceOneDestinationOperation.java]: Add documentation for how and when to use this

File: core/src/main/java/edu/wpi/grip/core/operations/opencv/templated/TwoSourceOneDestinationOperation.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.util.List;
+
+/**
+ * An Operation that takes two input values and produces one output value
+ * @param <T1>
+ * @param <T2>
+ * @param <R>
+ */
+public class TwoSourceOneDestinationOperation<T1, T2, R> implements Operation {
```
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/opencv/templated/TwoSourceOneDestinationOperation.java]: Improve the documentation. What are T1, T2, and R?

File: core/src/main/java/edu/wpi/grip/core/sockets/OutputSocketImpl.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+import com.thoughtworks.xstream.annotations.XStreamAlias;
+
+import edu.wpi.grip.core.events.SocketPreviewChangedEvent;
+
+/**
+ *
+ * @param <T>
+ */
+@XStreamAlias("grip:Output")
+public class OutputSocketImpl<T> extends SocketImpl<T> implements OutputSocket<T> {
```
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/sockets/OutputSocketImpl.java]: Document this. Even a basic copy/paste from `InputSocketImpl` would be okay.

File: core/src/main/java/edu/wpi/grip/core/util/Icons.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        return iconStream("/edu/wpi/grip/ui/icons/", name, type);
+    }
+
+    /**
+     * Gets an image stream for an icon.
+     *
+     * @param name the name of the icon
+     * @return a stream for the given icon, or {@code null} if no image by that name exists
+     */
+    public static InputStream iconStream(String name) {
```
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/util/Icons.java]: Improve the documentation for this -- it'll look for the image at `/edu/wpi/grip/ui/icons/{name}.png`

File: core/src/main/java/edu/wpi/grip/core/util/Icons.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+
+    /**
+     * Gets an image stream for an icon.
+     *
+     * @param name the name of the icon
+     * @param type the type of the icon (".png", ".jpg", etc.)
+     * @return a stream for the given icon, or {@code null} if no image by that name exists
+     */
+    public static InputStream iconStream(String name, String type) {
```
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/util/Icons.java]: Improve the documentation for this -- it'll look for the image at `/edu/wpi/grip/ui/icons/{name}.{type}`

File: core/src/main/java/edu/wpi/grip/core/OperationDescription.java [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+                .category(Category.MISCELLANEOUS)
+                .icon(null);
+    }
+
+    /**
+     * Builder class for {@code OperationDescription}
+     */
+    public static final class Builder {
+        private String name;
+        private String summary = "PLEASE PROVIDE A DESCRIPTION TO THE OPERATION DESCRIPTION!";
```
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/OperationDescription.java]: Why is this the default? The summary should either be empty (like it is right now) or `null` and throw an `NPE` when `build()` is called

File: buildSrc/src/main/java/edu/wpi/gripgenerator/FileParser.java [hunk 14]
```diff
@@ -127,69 +123,69 @@ public Expression getDefaultValue(String defaultValue) {
...
-                ).addDescription("Resize the image to the specified size."),
-                new DefinedMethod("rectangle", false,
-                        new DefinedParamType("Mat", DefinedParamType.DefinedParamDirection.INPUT_AND_OUTPUT),
-                        new DefinedParamType("Point")
-                ).addDescription("Draw a rectangle (outline or filled) on an image.")
-        ).setDirectionDefaults(DefinedParamType.DefinedParamDirection.OUTPUT, "dst")
-                .setIgnoreDefaults("dtype", "ddepth");
-        new OpenCVMethodVisitor(collection).visit(imgprocDeclaration, compilationUnits);
-        collection.generateCompilationUnits(collector, compilationUnits, operations);
+//        DefinedMethodCollection collection = new DefinedMethodCollection(baseClassName,
```
Comment (SamCarlberg) [inline buildSrc/src/main/java/edu/wpi/gripgenerator/FileParser.java]: Remove all these commented lines if they're not going to be uncommented.

File: core/src/main/java/edu/wpi/grip/core/OperationDescription.java (hunk 13 above)
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/OperationDescription.java]: Because in tests its fine, when creating operations its good for simple testing.
Otherwise, it should show up in the pallette blatantly obvious that it is a problem.

**Comment by SamCarlberg** [comment]
For some reason, most operations in the palette don't display their images.

File: core/src/main/java/edu/wpi/grip/core/operations/Operations.java [hunk 15]
```diff
@@ -20,82 +20,86 @@
...
+                new OperationMetaData(ValveOperation.DESCRIPTION, () -> new ValveOperation(isf, osf)),
+                new OperationMetaData(WatershedOperation.DESCRIPTION, () -> new WatershedOperation(isf, osf)),
+
+                // OpenCV operations
+                new OperationMetaData(MatFieldAccessor.DESCRIPTION, () -> new MatFieldAccessor(isf, osf)),
+                new OperationMetaData(MinMaxLoc.DESCRIPTION, () -> new MinMaxLoc(isf, osf)),
+                new OperationMetaData(NewPointOperation.DESCRIPTION, () -> new NewPointOperation(isf, osf)),
+                new OperationMetaData(NewSizeOperation.DESCRIPTION, () -> new NewSizeOperation(isf, osf)),
+
+                // Raw OpenCV operations
```
Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/Operations.java]: Remove this comment and associated newlines since OpenCV operations are in their own file

**Comment by SamCarlberg** [comment]
Missing operations:
```
ROSPublish ContoursReport
ROSPublish BlobsReport
ROSPublish LinesReport
Threshold Moving
CV Erode
```

File: core/src/main/java/edu/wpi/grip/core/Connection.java [hunk 16]
```diff
@@ -76,14 +76,14 @@ public void removeConnection(StepRemovedEvent e) {
         // Remove this connection if one of the steps it was connected to was removed
         for (OutputSocket socket : e.getStep().getOutputSockets()) {
             if (socket == this.outputSocket) {
-                this.eventBus.post(new ConnectionRemovedEvent(this));
+                this.remove();
```
Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/Connection.java]: Can you remove `this`?

File: core/src/main/java/edu/wpi/grip/core/Connection.java [hunk 17]
```diff
@@ -76,14 +76,14 @@ public void removeConnection(StepRemovedEvent e) {
...
                 return;
             }
         }
 
         for (InputSocket socket : e.getStep().getInputSockets()) {
             if (socket == this.inputSocket) {
-                this.eventBus.post(new ConnectionRemovedEvent(this));
+                this.remove();
```
Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/Connection.java]: Can you remove `this`?

File: core/src/main/java/edu/wpi/grip/core/Connection.java [hunk 18]
```diff
@@ -94,9 +94,16 @@ public void removeConnection(SourceRemovedEvent e) {
         // Remove this connection if it's from a source that was removed
         for (OutputSocket socket : e.getSource().getOutputSockets()) {
             if (socket == this.outputSocket) {
-                this.eventBus.post(new ConnectionRemovedEvent(this));
+                this.remove();
```
Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/Connection.java]: Can you remove `this`?

Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/Connection.java]: Actually, why not do something like this here:
```
if (e.getSource().getOutputSockets().contains(this.outputSocket)) {
    remove();
}
```

File: core/src/main/java/edu/wpi/grip/core/Connection.java (hunk 16 above)
Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/Connection.java]: See my note on line 97

File: core/src/main/java/edu/wpi/grip/core/Connection.java (hunk 17 above)
Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/Connection.java]: See my note on line 97

File: core/src/main/java/edu/wpi/grip/core/Palette.java [hunk 19]
```diff
@@ -27,13 +27,13 @@
         this.eventBus = eventBus;
```
Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/Palette.java]: Not part of this PR but does this class need the EventBus?

File: core/src/main/java/edu/wpi/grip/core/operations/composite/BlurOperation.java [hunk 20]
```diff
@@ -40,50 +52,41 @@ public String toString() {
...
-    public InputSocket<?>[] createInputSockets(EventBus eventBus) {
-        return new InputSocket<?>[]{
-                new InputSocket<>(eventBus, inputHint),
-                new InputSocket<>(eventBus, typeHint),
-                new InputSocket<>(eventBus, radiusHint),
-        };
+    public List<InputSocket> getInputSockets() {
+        return ImmutableList.of(
+                inputSocket,
+                typeSocket
```
Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/BlurOperation.java]: Add radius socket here

File: core/src/main/java/edu/wpi/grip/core/Source.java [hunk 21]
```diff
@@ -61,15 +62,15 @@ protected Source(ExceptionWitness.Factory exceptionWitnessFactory) {
      * @return @return An array of {@link OutputSocket}s for the outputs that the source produces.
      */
     public final ImmutableList<OutputSocket> getOutputSockets() {
-        final OutputSocket[] outputSockets = this.createOutputSockets();
+        final List<OutputSocket> outputSockets = this.createOutputSockets();
```
Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/Source.java]: Do we need the `this` here?

File: core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java [hunk 22]
```diff
@@ -31,16 +32,41 @@
...
+    private final InputSocket<Number> qualitySocket;
+
+    public PublishVideoOperation(InputSocket.Factory inputSocketFactory, OutputSocket.Factory outputSocketFactory) {
+        if (numSteps != 0) {
+            throw new IllegalStateException("Only one instance of PublishVideoOperation may exist");
+        }
+        this.inputSocket = inputSocketFactory.create(SocketHints.Inputs.createMatSocketHint("Image", false));
+        this.qualitySocket = inputSocketFactory.create(SocketHints.Inputs.createNumberSliderSocketHint("Quality", 80, 0, 100));
+        numSteps = numSteps + 1;
+        if (serverThread == null) {
```
Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java]: Why check to see if serverThread is null?  It should always be null in the constructor

File: core/src/main/java/edu/wpi/grip/core/operations/composite/ValveOperation.java [hunk 23]
```diff
@@ -1,63 +1,59 @@
...
+        this.inputSocket = linkedSocketHint.linkedInputSocket("Input");
+
+        this.outputSocket = linkedSocketHint.linkedOutputSocket("Output");
     }
 
     @Override
-    public InputSocket<?>[] createInputSockets(EventBus eventBus) {
-        throw new UnsupportedOperationException("This method should not be used");
+    public List<InputSocket> getInputSockets() {
+        return ImmutableList.of(switcherSocket, inputSocket);
```
Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/ValveOperation.java]: Style with new lines

File: core/src/main/java/edu/wpi/grip/core/operations/composite/ValveOperation.java [hunk 24]
```diff
@@ -1,63 +1,59 @@
...
     }
 
     @Override
-    public OutputSocket<?>[] createOutputSockets(EventBus eventBus) {
-        throw new UnsupportedOperationException("This method should not be used");
+    public List<OutputSocket> getOutputSockets() {
+        return ImmutableList.of(outputSocket);
```
Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/ValveOperation.java]: Style with new lines

File: core/src/main/java/edu/wpi/grip/core/operations/composite/ThresholdMoving.java [hunk 25]
```diff
@@ -1,64 +1,48 @@
...
+        lastImage = new Mat();
     }
 
     @Override
-    public InputSocket<?>[] createInputSockets(EventBus eventBus) {
-        return new InputSocket<?>[]{
-                new InputSocket<>(eventBus, SocketHints.Inputs.createMatSocketHint("image", false))
-        };
+    public List<InputSocket> getInputSockets() {
+        return ImmutableList.of(imageSocket);
```
Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/ThresholdMoving.java]: Style with new lines

File: core/src/main/java/edu/wpi/grip/core/operations/composite/ThresholdMoving.java [hunk 26]
```diff
@@ -1,64 +1,48 @@
...
     }
 
     @Override
-    public OutputSocket<?>[] createOutputSockets(EventBus eventBus) {
-        return new OutputSocket<?>[]{
-                new OutputSocket<>(eventBus, SocketHints.Outputs.createMatSocketHint("moved"))
-        };
+    public List<OutputSocket> getOutputSockets() {
+        return ImmutableList.of(outputSocket);
```
Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/ThresholdMoving.java]: Style with new lines

File: core/src/test/java/edu/wpi/grip/core/AdditionOperation.java [hunk 27]
```diff
@@ -1,42 +1,48 @@
...
+        a = isf.create(aHint);
+        b = isf.create(bHint);
+        c = osf.create(cHint);
     }
 
     @Override
-    public InputSocket[] createInputSockets(EventBus eventBus) {
-        return new InputSocket[]{new InputSocket<>(eventBus, aHint), new InputSocket<>(eventBus, bHint)};
+    public List<InputSocket> getInputSockets() {
+        return ImmutableList.of(a, b);
```
Comment (AustinShalit) [inline core/src/test/java/edu/wpi/grip/core/AdditionOperation.java]: Newline style

File: core/src/test/java/edu/wpi/grip/core/AdditionOperation.java [hunk 28]
```diff
@@ -1,42 +1,48 @@
...
     }
 
     @Override
-    public OutputSocket[] createOutputSockets(EventBus eventBus) {
-        return new OutputSocket[]{new OutputSocket<>(eventBus, cHint)};
+    public List<OutputSocket> getOutputSockets() {
+        return ImmutableList.of(c);
```
Comment (AustinShalit) [inline core/src/test/java/edu/wpi/grip/core/AdditionOperation.java]: Newline style

File: core/src/main/java/edu/wpi/grip/core/sources/MultiImageFileSource.java [hunk 29]
```diff
@@ -113,8 +119,8 @@ public String getName() {
     }
 
     @Override
-    protected OutputSocket[] createOutputSockets() {
-        return new OutputSocket[]{outputSocket};
+    protected List<OutputSocket> createOutputSockets() {
+        return ImmutableList.of(outputSocket);
```
Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/sources/MultiImageFileSource.java]: Newline style

File: core/src/main/java/edu/wpi/grip/core/sources/ImageFileSource.java [hunk 30]
```diff
@@ -93,8 +98,8 @@ public String getName() {
     }
 
     @Override
-    public OutputSocket[] createOutputSockets() {
-        return new OutputSocket[]{outputSocket};
+    public List<OutputSocket> createOutputSockets() {
+        return ImmutableList.of(outputSocket);
```
Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/sources/ImageFileSource.java]: Newline style

File: core/src/main/java/edu/wpi/grip/core/sources/CameraSource.java [hunk 31]
```diff
@@ -237,8 +242,8 @@ public String getName() {
     }
 
     @Override
-    public OutputSocket[] createOutputSockets() {
-        return new OutputSocket[]{frameOutputSocket, frameRateOutputSocket};
+    public List<OutputSocket> createOutputSockets() {
+        return ImmutableList.of(frameOutputSocket, frameRateOutputSocket);
```
Comment (AustinShalit) [inline core/src/main/java/edu/wpi/grip/core/sources/CameraSource.java]: Newline style

**Comment by AustinShalit** [comment]
Regarding the icon issue: InputStreams can only be read once.  Two ways to solve it:

1. Mark the stream before you read it and then reset it every time you are done reading the stream

2. Use the stream once to save the information to an Image object

**Comment by SamCarlberg** [comment]
Huh, I figured that the constructor would close the stream. Should just change the icons utility class to return the icons (and cache them in memory) instead of streams.

File: core/src/main/java/edu/wpi/grip/core/Palette.java (hunk 19 above)
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/Palette.java]: No. No it really doesn't.
